// @tenderly/hardhat-tenderly/dist/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContracts = void 0;
exports.getContracts = async (hre) => {
    const sourcePaths = await hre.run("compile:solidity:get-source-paths");
    const sourceNames = await hre.run("compile:solidity:get-source-names", {
        sourcePaths
    });
    const data = await hre.run("compile:solidity:get-dependency-graph", {
        sourceNames
    });
    const requestContracts = [];
    data._resolvedFiles.forEach((resolvedFile, _) => {
        var _a;
        const name = resolvedFile.sourceName
            .split("/")
            .slice(-1)[0]
            .split(".")[0];
        const contractToPush = {
            contractName: name,
            source: resolvedFile.content.rawContent,
            sourcePath: resolvedFile.sourceName,
            networks: {},
            compiler: {
                name: "solc",
                version: (_a = hre.config.solidity) === null || _a === void 0 ? void 0 : _a.compilers[0].version
            }
        };
        requestContracts.push(contractToPush);
    });
    return requestContracts;
};
