// @tenderly/hardhat-tenderly/dist/tenderly/TenderlyService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TenderlyService = exports.TENDERLY_RPC_BASE = exports.TENDERLY_DASHBOARD_BASE_URL = exports.TENDERLY_API_BASE_URL = void 0;
const index_1 = require("../index");
const TenderlyApiService_1 = require("./TenderlyApiService");
exports.TENDERLY_API_BASE_URL = "https://api.tenderly.co";
exports.TENDERLY_DASHBOARD_BASE_URL = "https://dashboard.tenderly.co";
// TODO(viktor): change this when we release rpc
exports.TENDERLY_RPC_BASE = "https://rpc.tenderly.co";
class TenderlyService {
    static async verifyContracts(request) {
        const tenderlyApi = TenderlyApiService_1.TenderlyApiService.configureInstance();
        try {
            const response = await tenderlyApi.post("/api/v1/account/me/verify-contracts", Object.assign({}, request));
            const responseData = response.data;
            let contract;
            console.log("Smart Contracts successfully verified");
            console.group();
            for (contract of responseData.contracts) {
                const contractLink = `${exports.TENDERLY_DASHBOARD_BASE_URL}/contract/${index_1.ReverseNetworkMap[contract.network_id]}/${contract.address}`;
                console.log(`Contract ${contract.address} verified. You can view the contract at ${contractLink}`);
            }
            console.groupEnd();
        }
        catch (error) {
            console.log(`Error in ${index_1.PluginName}: There was an error during the request. Contract verification failed`);
        }
    }
    static async pushContracts(request, tenderlyProject, username) {
        const tenderlyApi = TenderlyApiService_1.TenderlyApiService.configureInstance();
        try {
            await tenderlyApi.post(`/api/v1/account/${username}/project/${tenderlyProject}/contracts`, Object.assign({}, request));
            const dashLink = `${exports.TENDERLY_DASHBOARD_BASE_URL}/${username}/${tenderlyProject}/contracts`;
            console.log(`Successfully pushed Smart Contracts for project ${tenderlyProject}. You can view your contracts at ${dashLink}`);
        }
        catch (error) {
            console.log(`Error in ${index_1.PluginName}: There was an error during the request. Contract push failed`);
        }
    }
    static async verifyForkContracts(request, tenderlyProject, username, fork) {
        const tenderlyApi = TenderlyApiService_1.TenderlyApiService.configureTenderlyRPCInstance();
        try {
            const response = await tenderlyApi.post(`/account/${username}/project/${tenderlyProject}/fork/${fork}/verify`, Object.assign({}, request));
            const responseData = response.data;
            if (responseData.bytecode_mismatch_errors != null) {
                console.log(`Error in ${index_1.PluginName}: Bytecode mismatch detected. Contract verification failed`);
                return;
            }
            console.log("Smart Contracts successfully verified");
        }
        catch (error) {
            console.log(`Error in ${index_1.PluginName}: There was an error during the request. Contract verification failed`);
        }
    }
}
exports.TenderlyService = TenderlyService;
